craft_application/services/provider.py

```diff
@@ -0,0 +1,184 @@
+                # Ignore argument type until craft-providers accepts PurePosixPaths
+                # https://github.com/canonical/craft-providers/issues/315
+                target=self._app.managed_instance_project_path,  # type: ignore[arg-type]
+            )
+            emit.trace("Instance launched and working directory mounted")
+            yield instance
+
+    def get_base(
+        self,
+        base_name: bases.BaseName | tuple[str, str],
+        *,
+        instance_name: str,
+        **kwargs: bool | str | None,
+    ) -> craft_providers.Base:
+        """Get the base configuration from a base name.
+
+        :param base_name: The base to lookup.
+        :param instance_name: A name to assign to the instance.
+        :param kwargs: Additional keyword arguments are sent directly to the base.
+
+        This method should be overridden by a specific application if it intends to
+        use base names that don't align to a "distro:version" naming convention.
+        """
+        alias = bases.get_base_alias(base_name)
+        base_class = bases.get_base_from_alias(alias)
+        return base_class(
+            alias=alias,
+            hostname=instance_name,
+            snaps=self.snaps,
+            environment=self.environment,
+            **kwargs,  # type: ignore[arg-type]
+        )
+
+    def get_provider(self, name: str | None = None) -> craft_providers.Provider:
+        """Get the provider to use.
+
+        :param name: if set, uses the given provider name.
+
+        """
+        if self._provider is not None:
+            return self._provider
+        if self.is_managed:
+            raise CraftError("Cannot nest managed environments.")
+        if name is None:
+            emit.debug("Using default provider")
+            self._provider = self._get_default_provider()
+            return self._provider
+        emit.debug(f"Using provider {name!r}")
+        self._provider = self._get_provider_by_name(name)
+        return self._provider
+
+    def _get_default_provider(self) -> craft_providers.Provider:
+        """Get the default provider class to use for this application.
+
+        :returns: An instance of the default Provider class.
+        """
+        if sys.platform == "linux":
+            emit.trace("Linux detected, using LXD.")
+            return self._get_lxd_provider()
+        emit.trace("Non-linux platform. Using Multipass.")
+        return self._get_multipass_provider()
+
+    def _get_provider_by_name(self, name: str) -> craft_providers.Provider:
+        """Get a provider by its name."""
+        if name == "lxd":
+            return self._get_lxd_provider()
+        if name == "multipass":
+            return self._get_multipass_provider()
+        raise RuntimeError(f"Unknown provider: {name!r}")
+
+    def _get_lxd_provider(self) -> LXDProvider:
+        """Get the LXD provider for this manager."""
+        # TODO: Replace this deprecated function.
```

Comment:
See https://github.com/canonical/craft-providers/pull/318

I don't think you need to call this here, the replacement functions are already called from [`LXDProvider`](https://github.com/canonical/craft-providers/blob/832784819045ca5b18240d74eacf93f5bb482a7c/craft_providers/lxd/lxd_provider.py#L126-L127)

The primary reason I left the function and added a deprecation was for `lpci`, since they haven't switched to the LXDProvider class.

Reply:
Thanks! This code was directly from the old prototype. I don't remember exactly why I added it, but at some point I needed it. I've removed that.